perf_test/sparse/KokkosSparse_gs.cpp

```diff
@@ -187,92 +188,170 @@ void runGS(const GS_Parameters& params) {
   }
 
-  double applyLaunchTime = timer.seconds();
+  double blockExtractionTime = 0, symbolicLaunchTimeTotal = 0,
+         symbolicComputeTimeTotal = 0, numericLaunchTimeTotal = 0,
+         numericComputeTimeTotal = 0, applyLaunchTimeTotal = 0,
+         applyComputeTimeTotal = 0;
+
   timer.reset();
+  KokkosSparse::Impl::kk_extract_diagonal_blocks_crsmatrix_sequential(A,
+                                                                      DiagBlks);
   Kokkos::fence();
-  double applyComputeTime = timer.seconds();
-  timer.reset();
-  kh.destroy_gs_handle();
-  // Now, compute the 2-norm of residual
-  scalar_view_t res("Ax-b", nrows);
-  Kokkos::deep_copy(res, b);
-  scalar_t alpha = Kokkos::reduction_identity<scalar_t>::prod();
-  scalar_t beta  = -alpha;
-  KokkosSparse::spmv<scalar_t, crsMat_t, scalar_view_t, scalar_t,
-                     scalar_view_t>("N", alpha, A, x, beta, res);
-  double resnorm = KokkosBlas::nrm2(res);
-  std::cout << "\n*** Symbolic launch time: " << symbolicLaunchTime << '\n';
-  std::cout << "\n*** Symbolic compute time: " << symbolicComputeTime << '\n';
-  std::cout << "\n*** Numeric launch time: " << numericLaunchTime << '\n';
-  std::cout << "\n*** Numeric compute time: " << numericComputeTime << '\n';
-  std::cout << "\n*** Apply launch time: " << applyLaunchTime << '\n';
-  std::cout << "\n*** Apply compute time: " << applyComputeTime << '\n';
-  // note: this still works if the solution diverges
-  std::cout << "Relative res norm: " << resnorm / bnorm << '\n';
+  blockExtractionTime = timer.seconds();
+
+  for (int i = 0; i < params.nstreams; i++) {
+    auto blk_A     = DiagBlks[i];
+    auto blk_nrows = blk_A.numRows();
+    auto blk_ncols = blk_A.numCols();
+    if (blk_nrows != blk_ncols) {
+      cout << "ERROR: Gauss-Seidel only works for square matrices\n";
+      Kokkos::finalize();
+      exit(1);
+    }
+    b[i] = scalar_view_t("b[" + std::to_string(i) + "]", blk_nrows);
+    x[i] = scalar_view_t("x[" + std::to_string(i) + "]", blk_nrows);
+    {
+      srand(54321);
+      auto bhost = Kokkos::create_mirror_view(b[i]);
+      for (lno_t row_id = 0; row_id < blk_nrows; row_id++) {
+        bhost(row_id) = 10.0 * rand() / RAND_MAX - 5.0;
+      }
+      Kokkos::deep_copy(instances[i], b[i], bhost);
+    }
+    double bnorm = KokkosBlas::nrm2(instances[i], b[i]);
+    // cluster size of 1 is standard multicolor GS
+    if (params.algo == GS_DEFAULT) {
+      kh[i].create_gs_handle(instances[i], params.nstreams);
+      kh[i].get_point_gs_handle()->set_long_row_threshold(
+          params.longRowThreshold);
+    } else if (params.algo == GS_CLUSTER) {
+      kh[i].create_gs_handle(params.coarse_algo, params.cluster_size);
+    } else {
+      kh[i].create_gs_handle(params.algo);
+      if (params.algo == GS_TWOSTAGE)
+        kh[i].set_gs_twostage(!params.classic, blk_nrows);
+    }
+    timer.reset();
+    KokkosSparse::Experimental::gauss_seidel_symbolic(
```

Comment:
Shouldn't each step (symbolic, numeric, apply) be in its own loop over streams, followed by a single fence to finish computation for all the streams? Since this is launching each step and then doing a ``Kokkos::fence``, there's no concurrency between streams